### PR TITLE
Forbid borrows and unsized types from being used as the type of a const generic under `adt_const_params`

### LINKED ISSUE #112124: `adt_const_params` allows impls of `ConstParamTy` with non-ConstParamTy pointees
I tried this code:

```rust
#![feature(adt_const_params)]

use core::marker::ConstParamTy;

struct Foo;

impl ConstParamTy for &'static Foo {}
```

I expected to see this happen: it should error

Instead, this happened: it compilers

### Meta

`playground version`:
```
Nightly channel

Build using the Nightly version: 1.72.0-nightly

(2023-05-29 498553fc04f6a3fdc534)
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

### LINKED ISSUE #112125: `adt_const_params` does not explicitly check for a union `ConstParamTy` impl
I am not sure there is actually a piece of code that is incorrectly compiled since you cannot manually impl `StructuralEq` and `StructuralPartialEq` and you also cant `derive(Eq, PartialEq)` on a union. regardless there is currently no check that a manual `impl ConstParamTy for MyType {` is not applied to a union. There should be one to ensure that we do not accidentally allow usage of unions in const param tys.

### Meta

`rustc --version --verbose`:
```
Nightly channel

Build using the Nightly version: 1.72.0-nightly

(2023-05-29 498553fc04f6a3fdc534)
```

<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

### LINKED ISSUE #112219: sizedness of const parameter types
currently there is no restriction on const parameter types having to be sized, although it is impossible to provide any arguments to a parameter such as `const N: [u8]` because type checking `{ [1; 2] as [u8] }` will fail as `[u8]` is not `Sized`. We should figure out if we want to find some way to allow !Sized const generics to work or if we want to explicitly forbid them (although if we do we should not add `Sized` as a supertrait of `ConstParamTy` as that would make it impossible to ever allow `!Sized` const parameter types)

### LINKED ISSUE #120961: How should const generics with references work around pointer identity and padding?
Const generics currently completely deconstruct the value into a valtree and then reconstruct it, which destroys pointer identity. As long as this is just about purely immutable `const` values we might be able to argue that they "don't have pointer identity" (though it is entirely unclear to me what exactly that should mean). But if consts can [point to statics](https://github.com/rust-lang/rust/issues/119618), that doesn't really work any more.

Example:
```rust
#![feature(const_refs_to_static)]
#![feature(adt_const_params)]

static FOO: usize = 42;
const BAR: &usize = &FOO;
fn foo<const X: &'static usize>() {
    if std::ptr::eq(X, &FOO) {
        println!("activating special mode");
    }
}

fn main() {
    foo::<BAR>();
}

```
Arguably this should print, but [currently it does not](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=19072469cd8ddab79da2252616fe8684).

Cc https://github.com/rust-lang/rust/issues/95174, https://github.com/rust-lang/rust/issues/119618

Similarly, this example demonstrates that padding is lost when passing an `&(u8, u16)`  as a const generic, which is not how passing a reference usually behaves:
```rust
#![feature(const_refs_to_static)]
#![feature(adt_const_params)]

static FOO: u32 = 0x01020304;
const BAR: &(u8, u16) = unsafe { std::mem::transmute(&FOO) };

fn foo<const X: &'static (u8, u16)>() {
    let val: &u32 = unsafe { std::mem::transmute(X) };
    println!("{val:#x}"); // prints 0x1020004
}

fn main() {
    foo::<BAR>();
}
```

### PR DESCRIPTION
**Pull request by BoxyUwU**
Fixes #112219
Fixes #112124
Fixes #112125

### Motivation

Currently the `adt_const_params` feature allows writing `Foo<const N: [u8]>` this is entirely useless as it is not possible to write an expression which evaluates to a type that is not `Sized`. In order to actually use unsized types in const generics they are typically written as `const N: &[u8]` which *is* possible to provide a value of.

Unfortunately allowing the types of const parameters to contain references is non trivial (#120961) as it introduces a number of difficult questions about how equality of references in the type system should behave. References in the types of const generics is largely only useful for using unsized types in const generics.

This PR introduces a new feature gate `unsized_const_parameters` and moves support for `const N: [u8]` and `const N: &...` from `adt_const_params` into it. The goal here hopefully is to experiment with allowing `const N: [u8]` to work without references and then eventually completely forbid references in const generics. 

Splitting this out into a new feature gate means that stabilization of `adt_const_params` does not have to resolve #120961 which is the only remaining "big" blocker for the feature. Remaining issues after this are a few ICEs and naming bikeshed for `ConstParamTy`.

### Implementation

The implementation is slightly subtle here as we would like to ensure that a stabilization of `adt_const_params` is forwards compatible with any outcome of `unsized_const_parameters`. This is inherently tricky as we do not support unstable trait implementations and we determine whether a type is valid as the type of a const parameter via a trait bound.

There are a few constraints here:
- We would like to *allow for the possibility* of adding a `Sized` supertrait to `ConstParamTy` in the event that we wind up opting to not support unsized types and instead requiring people to write the 'sized version', e.g. `const N: [u8; M]` instead of `const N: [u8]`.
- Crates should be able to enable `unsized_const_parameters` and write trait implementations of `ConstParamTy` for `!Sized` types without downstream crates that only enable `adt_const_params` being able to observe this (required for std to be able to `impl<T> ConstParamTy for [T]`

Ultimately the way this is accomplished is via having two traits (sad), `ConstParamTy` and `UnsizedConstParamTy`. Depending on whether `unsized_const_parameters` is enabled or not we change which trait is used to check whether a type is allowed to be a const parameter.

Long term (when stabilizing `UnsizedConstParamTy`) it should be possible to completely merge these traits (and derive macros), only having a single `trait ConstParamTy` and `macro ConstParamTy`.

Under `adt_const_params` it is now illegal to directly refer to `ConstParamTy` it is only used as an internal impl detail by `derive(ConstParamTy)` and checking const parameters are well formed. This is necessary in order to ensure forwards compatibility with all possible future directions for `feature(unsized_const_parameters)`.

Generally the intuition here should be that `ConstParamTy` is the stable trait that everything uses, and `UnsizedConstParamTy` is that plus unstable implementations (well, I suppose `ConstParamTy` isn't stable yet :P).

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/misc.rs [hunk 1]
```diff
@@ -89,33 +92,113 @@ pub fn type_allowed_to_implement_const_param_ty<'tcx>(
...
+    if !tcx.features().unsized_const_parameters {
+        // Check for sizedness before recursing into ADT fields so that if someone tries to write:
+        // ```rust
+        //  #[derive(ConstParamTy)]
+        //  struct Foo([u8])
+        // ```
+        // They are told that const parameter types must be sized, instead of it saying that
+        // the trait implementation `[u8]: ConstParamTy` is not satisfied.
+        let infcx = tcx.infer_ctxt().build();
+        let ocx = traits::ObligationCtxt::new_with_diagnostics(&infcx);
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/misc.rs:104]:
```suggestion
        let ocx = traits::ObligationCtxt::new(&infcx);
```

If you're not actually extracting a `FulfillmentError` here, it's preferable not to generate them.

**Review by compiler-errors — APPROVED** [review]
reviewed the compiler part, only one nit

File: compiler/rustc_span/src/symbol.rs [hunk 2]
```diff
@@ -2000,6 +2002,8 @@ symbols! {
         unsafe_no_drop_flag,
         unsafe_pin_internals,
         unsize,
+        unsized_const_param_ty,
+        unsized_const_parameters,
```
Comment (compiler-errors) [inline compiler/rustc_span/src/symbol.rs:2006]:
```suggestion
        unsized_const_params,
```

what's up w the inconsistent naming?

Comment (BoxyUwU) [reply compiler/rustc_span/src/symbol.rs:2006]: no real reason, just figured it was less important for a feature gate to be short than a trait/derive that you'd write a bunch

Comment (compiler-errors) [reply compiler/rustc_span/src/symbol.rs:2006]: it's less that it needs to be short or long, it's just that it's pretty inconsistent with the other feature gate that users are likely familiar w/ if they reach for this one. i'd prefer if we change it

Comment (BoxyUwU) [reply compiler/rustc_span/src/symbol.rs:2006]: yeah it makes sense to keep them the same as eachother

**Comment by BoxyUwU** [comment]
Actually the diagnostics are fine, they could probably be better in some edge cases but that's pre existing and would much rather handle in a follow up PR than make this any bigger.

**Comment by BoxyUwU** [comment]
Unless CI decides to fail after a rebase this should be ready now lol

**Review by compiler-errors — APPROVED** [review]

**Comment by lqd** [comment]
The `deep-vector` results are noise, and returned to normal [in the next merged PR](https://github.com/rust-lang/rust/pull/127993#issuecomment-2241599216).

**Pull request merged**